source/utilities.tex

```diff
@@ -15469,6 +15999,201 @@
 \tcode{i < size_ ?\ data_[i] :\ basic_format_arg<Context>()}.
 \end{itemdescr}
 
+\rSec2[format.tuple]{Tuple formatter}
+
+\pnum
+For each of \tcode{pair} and \tcode{tuple},
+the library provides the following formatter specialization
+where \placeholder{tuple-type} is the name of the template:
+
+\indexlibraryglobal{formatter}%
+\begin{codeblock}
+namespace std {
+template<class charT, @\libconcept{formattable}@<charT>... Ts>
+  struct formatter<tuple-type<Ts...>, charT> {
+  private:
+    tuple<formatter<remove_cvref_t<Ts>, charT>...> @\exposid{underlying_}@;               // \expos
+    basic_string_view<charT> @\exposid{separator_}@ = @\exposid{STATICALLY-WIDEN}@<charT>(", ");      // \expos
+    basic_string_view<charT> @\exposid{opening-bracket_}@ = @\exposid{STATICALLY-WIDEN}@<charT>("("); // \expos
+    basic_string_view<charT> @\exposid{closing-bracket_}@ = @\exposid{STATICALLY-WIDEN}@<charT>(")"); // \expos
+
+  public:
+    constexpr void set_separator(basic_string_view<charT> sep);
+    constexpr void set_brackets(basic_string_view<charT> opening,
+	                        basic_string_view<charT> closing);
+
+    template<class ParseContext>
+      constexpr typename ParseContext::iterator
+        parse(ParseContext& ctx);
+
+    template<class FormatContext>
+      typename FormatContext::iterator
+        format(@\seebelow@& elems, FormatContext& ctx) const;
+  };
+}
+\end{codeblock}
+
+\pnum
+The \tcode{parse} member functions of these formatters
+interpret the format specification as
+a \fmtgrammarterm{tuple-format-spec} according to the following syntax:
+
+\begin{ncbnf}
+\fmtnontermdef{tuple-format-spec}\br
+    \opt{tuple-fill-and-align} \opt{width} \opt{tuple-type}
+\end{ncbnf}
+
+\begin{ncbnf}
+\fmtnontermdef{tuple-fill-and-align}\br
+    \opt{tuple-fill} align
+\end{ncbnf}
+
+\begin{ncbnf}
+\fmtnontermdef{tuple-fill}\br
+    \textnormal{any character other than} \terminal{\{} \textnormal{or} \terminal{\}} \textnormal{or} \terminal{:}
+\end{ncbnf}
+
+\begin{ncbnf}
+\fmtnontermdef{tuple-type}\br
```

Comment:
exactly. Should I rename one of them?

Reply:
Renamed grammarterms tuple-type to tuple-opt and range-type to range-opt

Reply:
Hi! I'd rather keep the _`tuple-type`_ and _`range-type`_ names for the grammar in the format specifiers - they're thematically similar to the _`type`_ for the default formatters. Can we instead change the _`tuple-type`_ that is used as a placeholder for `pair` or `tuple` to something like _`tuple-like-class`_ or even just... _`pair-or-tuple`_? 

Honestly now that I write that, _`pair-or-tuple`_ for sure.

Reply:
I think (and the paper author agrees) it would be better to undo your changes to the grammar terms and rename the `tuple-type` placeholder in the `formatter` specialization to be `tuple-like-type` or `tuple-like-class` or something like that.

Reply:
Oh our comments crossed in the tubes. I like Barry's second suggestion, _pair-or-tuple_

Reply:
done - renamed to pair-or-tuple